presto-main/src/main/java/com/facebook/presto/matching/MatchingEngine.java

```diff
@@ -63,12 +66,26 @@ private MatchingEngine(ListMultimap<Class, T> matchablesByClass)
             if (pattern instanceof Pattern.TypeOf) {
                 matchablesByClass.put(((Pattern.TypeOf) pattern).getType(), matchable);
             }
+            else if (pattern instanceof Pattern.V2Adapter) {
+                Pattern.V2Adapter adapter = (Pattern.V2Adapter) pattern;
+                com.facebook.presto.matching.v2.Pattern<?> firstPattern = getFirstPattern(adapter.getPattern());
+                checkArgument(firstPattern instanceof TypeOfPattern,
+                        "Unsupported v2 pattern, expected a TypeOfPattern, got " + firstPattern);
```

Comment:
It looks like there is missing code here, what are you going to do with `firstParent`